packages/react-reconciler/src/ReactFiberReconciler.js

```diff
@@ -276,7 +277,16 @@ export function createContainer(
   isConcurrent: boolean,
   hydrate: boolean,
 ): OpaqueRoot {
-  return createFiberRoot(containerInfo, isConcurrent, hydrate);
+  const fiberRoot = createFiberRoot(containerInfo, isConcurrent, hydrate);
+  if (__DEV__) {
+    // jest isn't a 'global', it's just exposed to tests via a wrapped function
+    // further, this isn't a test file, so flow doesn't recognize the symbol. So...
+    // $FlowExpectedError - because requirements don't give a damn about your type sigs.
+    if ('undefined' !== typeof jest) {
+      warnIfNotScopedWithMatchingAct(fiberRoot.current);
```

Comment:
Why does this warning fire but not the other one?

Reply:
I don't understand the question? In case a create and update are both wrapped by the wrong act(), then this warning will fire twice. But as explained in the description, the second one is less likely to fire, especially when async. We fire warnings for every violation.

Reply:
`createContainer` isn't the right place for it because it's not an update. It's often accompanied by an update  (`ReactDOM.render` and `TestRenderer.create` call both `createContainer` and `updateContainer` one after the other) but our newer APIs have shifted to a model where you create a root before scheduling updates on in, e.g. `ReactDOM.createRoot`, which creates a root but doesn't schedule any work.

But I think there's a higher level problem...

> But as explained in the description, the second one is less likely to fire, especially when async. We fire warnings for every violation.

I did read your description, and I think I understood most of it, but I don't know if I agree with your root cause analysis.

Is it correct to rephrase the problem like this? APIs that schedule an update but existed before hooks — `this.setState`, `ReactDOM.render`, `TestRenderer.create` — don't trigger a warning if you don't wrap them in `act`, because we didn't want to start firing warnings in old tests. However, these types of updates could still result in passive effects, and if you don't wrap them in `act`, the effects will silently remain unflushed without a warning.

So it seems to me that the solution is to warn if a passive effect is scheduled outside of `act`.

Reply:
I don't know about the suggested solution. That's a bit controversial.

However, this is clearly not the right place for this warning since creating a root doesn't mean I will render it yet. It just means I have a reference to it for when I will.

Reply:
- How about in updateContainer then?
- I could also add it explicitly for each renderer's 'render' - so ReactDOM.render(), root.render(), TestRenderer.create()/.update(), etc. This will also mean I'll have to export the warning on the reconciler.

Reply:
Let's continue discussion here: https://github.com/facebook/react/issues/15472

Reply:
I updated the PR to do the check inside updateContainer(), looks good

Reply:
also a PR for when effects get queued outside an act https://github.com/facebook/react/pull/15631